crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -141,12 +143,252 @@ pub unsafe trait SystemParam: Sized {
+        change_tick: Tick,
+    ) -> Self::Item<'world, 'state> {
+        fn handler<T>(err: impl Error) -> T {
+            panic!("{err}");
+        }
+        T::get_param(state, system_meta, world, change_tick).unwrap_or_else(handler)
+    }
+}
+
+unsafe impl<T: FallibleSystemParam> SystemParam for Option<T> {
+    type State = T::State;
+
+    type Item<'world, 'state> = Option<T::Item<'world, 'state>>;
+
+    fn init_state(world: &mut World, system_meta: &mut SystemMeta) -> Self::State {
+        T::init_state(world, system_meta)
+    }
+
+    fn new_archetype(state: &mut Self::State, archetype: &Archetype, system_meta: &mut SystemMeta) {
+        T::new_archetype(state, archetype, system_meta);
+    }
+
+    fn apply(state: &mut Self::State, system_meta: &SystemMeta, world: &mut World) {
+        T::apply(state, system_meta, world);
+    }
+
+    unsafe fn get_param<'world, 'state>(
+        state: &'state mut Self::State,
+        system_meta: &SystemMeta,
+        world: &'world World,
+        change_tick: Tick,
+    ) -> Self::Item<'world, 'state> {
+        T::get_param(state, system_meta, world, change_tick).ok()
+    }
+}
+
+unsafe impl<T: FallibleSystemParam> SystemParam for Result<T, T::Error> {
+    type State = T::State;
+
+    type Item<'world, 'state> = Result<T::Item<'world, 'state>, T::Error>;
+
+    fn init_state(world: &mut World, system_meta: &mut SystemMeta) -> Self::State {
+        T::init_state(world, system_meta)
+    }
+
+    fn new_archetype(state: &mut Self::State, archetype: &Archetype, system_meta: &mut SystemMeta) {
+        T::new_archetype(state, archetype, system_meta);
+    }
+
+    fn apply(state: &mut Self::State, system_meta: &SystemMeta, world: &mut World) {
+        T::apply(state, system_meta, world);
+    }
+
+    unsafe fn get_param<'world, 'state>(
+        state: &'state mut Self::State,
+        system_meta: &SystemMeta,
+        world: &'world World,
+        change_tick: Tick,
+    ) -> Self::Item<'world, 'state> {
+        T::get_param(state, system_meta, world, change_tick)
+    }
+}
+
+pub struct SystemParamError<T: FallibleSystemParam> {
+    system_name: Cow<'static, str>,
+    _marker: PhantomData<T>,
+}
+
+impl<T: FallibleSystemParam> SystemParamError<T> {
+    pub fn new(system_name: Cow<'static, str>) -> Self {
+        Self {
+            system_name,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<T: FallibleSystemParam> std::fmt::Debug for SystemParamError<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "SystemParamError<{}> {{ system_name: \"{}\" }}",
+            std::any::type_name::<T>(),
+            self.system_name
+        )
+    }
+}
+
+impl<T: FallibleSystemParam> std::fmt::Display for SystemParamError<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "`{}` requested by `{}` does not exist",
```

Comment:
This message seems a bit too specific to be used as a generic error message -- not all system param errors will be because the param does not exist. Maybe the error type could use a more descriptive name, something like `MissingSystemParam`?